### PR TITLE
Upgrade async-profiler to 3.0

### LINKED ISSUE #173: Migrate Inferred spans to async-profiler 3.x
Would be a pre-requisite to contribute this upstream

### PR DESCRIPTION
**Pull request by JonasKunz**
Closes #173 .

### CODE REVIEW

**Comment by JonasKunz** [comment]
I've also done manual testing using the elastiflix app and can confirm that inferred spans are still functional.

File: inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/BufferedFile.java [hunk 1]
```diff
@@ -110,6 +111,88 @@ public void skip(int bytesToSkip) {
...
+  }
+
+  /**
+   * @param output the buffer to place the string intro
+   * @return false, if the string to read is null, true otherwise
+   */
+  @Nullable
+  public boolean readString(StringBuilder output) throws IOException {
+    byte encoding = get();
+    if (encoding == 0) { // 0 encoding represents a null string
```
Comment (SylvainJuge) [inline inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/BufferedFile.java:125]: [nit] maybe use constant to help document this.

File: inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/BufferedFile.java [hunk 2]
```diff
@@ -110,6 +111,88 @@ public void skip(int bytesToSkip) {
...
+      return false;
+    }
+    readOrSkipString(encoding, output);
+    return true;
+  }
+
+  @Nullable
+  public String readString() throws IOException {
+    byte encoding = get();
+    if (encoding == 0) { // 0 encoding represents a null string
```
Comment (SylvainJuge) [inline inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/BufferedFile.java:135]: [nit] maybe use constant to help document this.

File: inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/BufferedFile.java [hunk 3]
```diff
@@ -110,6 +111,88 @@ public void skip(int bytesToSkip) {
...
+      return null;
+    }
+    if (encoding == 1) { // 1 encoding represents an empty string
```
Comment (SylvainJuge) [inline inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/BufferedFile.java:138]: [nit] maybe use constant to help document this.

File: inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/BufferedFile.java [hunk 4]
```diff
@@ -110,6 +111,88 @@ public void skip(int bytesToSkip) {
...
+        return;
+      case 3: // UTF8-encoded byte array
+        readOrSkipUtf8(output);
+        return;
+      case 4: // char array
+        throw new IllegalStateException("Char-array encoding is not supported by the parser yet");
+      case 5: // LATIN1-encoded byte array
+        if (output != null) {
+          throw new IllegalStateException("Reading LATIN1 encoded string is not supported");
+        }
```
Comment (SylvainJuge) [inline inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/BufferedFile.java:165]: [nit] maybe use constants to help document this.

File: inferred-spans/src/test/java/co/elastic/otel/profiler/asyncprofiler/JfrParserTest.java [hunk 5]
```diff
@@ -36,9 +36,9 @@ class JfrParserTest {
 
   @Test
   void name() throws Exception {
-    // using the smallest prime number possible for the buffer
-    // should trigger most edge cases in the buffer being exhausted
-    JfrParser jfrParser = new JfrParser(ByteBuffer.allocate(113), ByteBuffer.allocate(113));
+    // Using a small buffer, but big enough to fit the largest string in the JFR file to test edge
+    // cases
+    JfrParser jfrParser = new JfrParser(ByteBuffer.allocate(368), ByteBuffer.allocate(368));
```
Comment (SylvainJuge) [inline inferred-spans/src/test/java/co/elastic/otel/profiler/asyncprofiler/JfrParserTest.java:41]: [minor] where does this magic value comes from ? how to pick an appropriate value and maintain this in the future ?

File: inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/JfrParser.java [hunk 6]
```diff
@@ -105,169 +105,195 @@ public void parse(
...
     bufferedFile.setFile(file);
     long fileSize = bufferedFile.size();
-    if (fileSize < 16) {
+
+    int chunkSize = readChunk(0);
+    if (chunkSize < fileSize) {
       throw new IllegalStateException(
-          "Unexpected sampling profiler error, everything else should work as expected. "
-              + "Please report to us with as many details, including OS and JVM details.");
+          "This implementation does not support reading JFR files containing multiple chunks");
```
Comment (SylvainJuge) [inline inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/JfrParser.java:112]: [question] does it means we could make it work (probably more efficiently) with >1 chunk per file in the future ?

File: inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/JfrParser.java [hunk 7]
```diff
@@ -470,19 +477,23 @@ public interface StackTraceConsumer {
   private interface EventTypeId {
     int EVENT_METADATA = 0;
     int EVENT_CHECKPOINT = 1;
-    int EVENT_RECORDING = 10;
-    int EVENT_EXECUTION_SAMPLE = 20;
+
+    // The following event types actually are defined in thhe metadata of the JFR file itself
```
Comment (SylvainJuge) [inline inferred-spans/src/main/java/co/elastic/otel/profiler/asyncprofiler/JfrParser.java:481]:
```suggestion
    // The following event types actually are defined in the metadata of the JFR file itself
```

**Review by SylvainJuge — APPROVED** [review]
Nice work ! As usual I only have minor comments and questions.

File: inferred-spans/src/test/java/co/elastic/otel/profiler/asyncprofiler/JfrParserTest.java (hunk 5 above)
Comment (JonasKunz) [reply inferred-spans/src/test/java/co/elastic/otel/profiler/asyncprofiler/JfrParserTest.java:41]: It is just a small value to trigger the edge cases in `BufferdFile`, like before.
In the new JFR-file the longest string was `368` characters, which therefore is the lower bound to not crash the parsing.
